File: address-geocode-us-python/REST/agus_response.py
from dataclasses import dataclass
from typing import Optional, List


@dataclass
class GetDistanceToWaterInput:
    Latitude: Optional[float] = None
    Longitude: Optional[float] = None
    LicenseKey: Optional[str] = None
    IsLive: bool = True
    TimeoutSeconds: int = 15

    def __str__(self) -> str:
        return (f"GetDistanceToWaterInput: Latitude={self.Latitude}, Longitude={self.Longitude}, "
                f"LicenseKey={self.LicenseKey}, IsLive={self.IsLive}, TimeoutSeconds={self.TimeoutSeconds}")


@dataclass
class InformationComponent:
    Name: Optional[str] = None
    Value: Optional[str] = None

    def __str__(self) -> str:
        return f"InformationComponent: Name={self.Name}, Value={self.Value}"


@dataclass
class Error:
    Desc: Optional[str] = None
    Number: Optional[str] = None
    Location: Optional[str] = None

    def __str__(self) -> str:
        return (f"Error: Desc={self.Desc}, Number={self.Number}, Location={self.Location}")


@dataclass
class GetBestMatchV4Response:
    Level: Optional[str] = None
    LevelDescription: Optional[str] = None
    Latitude: Optional[float] = None
    Longitude: Optional[float] = None
    Zip: Optional[str] = None
    InformationComponents: Optional[List['InformationComponent']] = None
    Error: Optional['Error'] = None

    def __post_init__(self):
        if self.InformationComponents is None:
            self.InformationComponents = []

    def __str__(self) -> str:
        components_string = ', '.join(str(component) for component in self.InformationComponents) if self.InformationComponents else 'None'
        error = str(self.Error) if self.Error else 'None'
        return (f"GetBestMatchV4Response: Level={self.Level}, LevelDescription={self.LevelDescription}, "
                f"Latitude={self.Latitude}, Longitude={self.Longitude}, Zip={self.Zip}, "
                f"InformationComponents=[{components_string}], Error={error}")


@dataclass
class GetDistanceToWaterResponse:
    DistanceToWater: Optional[float] = None
    Latitude: Optional[float] = None
    Longitude: Optional[float] = None
    WaterLat: Optional[float] = None
    WaterLon: Optional[float] = None
    Error: Optional['Error'] = None

    def __str__(self) -> str:
        error = str(self.Error) if self.Error else 'None'
        return (f"GetDistanceToWaterResponse: DistanceToWater={self.DistanceToWater}, Latitude={self.Latitude}, "
                f"Longitude={self.Longitude}, WaterLat={self.WaterLat}, WaterLon={self.WaterLon}, Error={error}")


@dataclass
class GetReverseLocationResponse:
    Address: Optional[str] = None
    City: Optional[str] = None
    State: Optional[str] = None
    Zip: Optional[str] = None
    County: Optional[str] = None
    Error: Optional['Error'] = None

    def __str__(self) -> str:
        error = str(self.Error) if self.Error else 'None'
        return (f"GetReverseLocationResponse: Address={self.Address}, City={self.City}, State={self.State}, "
                f"Zip={self.Zip}, County={self.County}, Error={error}")

File: address-geocode-us-nodejs/REST/agus_response.js
export class GetDistanceToWaterInput {
    constructor(data = {}) {
        this.Latitude = data.Latitude;
        this.Longitude = data.Longitude;
        this.LicenseKey = data.LicenseKey;
        this.IsLive = data.IsLive !== undefined ? data.IsLive : true;
        this.TimeoutSeconds = data.TimeoutSeconds !== undefined ? data.TimeoutSeconds : 15;
    }

    toString() {
        return `GetDistanceToWaterInput: Latitude = ${this.Latitude}, Longitude = ${this.Longitude}, LicenseKey = ${this.LicenseKey}, IsLive = ${this.IsLive}, TimeoutSeconds = ${this.TimeoutSeconds}`;
    }
}

/**
 * Information Component for API responses.
 */
export class InformationComponent {
    constructor(data = {}) {
        this.Name = data.Name;
        this.Value = data.Value;
    }

    toString() {
        return `Name = ${this.Name}, Value = ${this.Value}`;
    }
}

/**
 * Error object for API responses.
 */
export class Error {
    constructor(data = {}) {
        this.Desc = data.Desc;
        this.Number = data.Number;
        this.Location = data.Location;
    }

    toString() {
        return `Error: Desc = ${this.Desc}, Number = ${this.Number}, Location = ${this.Location}`;
    }
}

/**
 * Response from GetBestMatch_V4 API.
 */
export class GetBestMatchV4Response {
    constructor(data = {}) {
        this.Level = data.Level;
        this.LevelDescription = data.LevelDescription;
        this.Latitude = data.Latitude;
        this.Longitude = data.Longitude;
        this.Zip = data.Zip;
        this.InformationComponents = (data.InformationComponents || []).map(component => new InformationComponent(component));
        this.Error = data.Error ? new Error(data.Error) : null;
    }

    toString() {
        const componentsString = this.InformationComponents.length
            ? this.InformationComponents.map(component => component.toString()).join(', ')
            : 'null';
        return `GetBestMatchV4Response: Level = ${this.Level}, LevelDescription = ${this.LevelDescription}, Latitude = ${this.Latitude}, Longitude = ${this.Longitude}, Zip = ${this.Zip}, InformationComponents = [${componentsString}], Error = ${this.Error ? this.Error.toString() : 'null'}`;
    }
}

/**
 * Response from GetDistanceToWater API, containing the estimated distance to the nearest saltwater
 * and the coordinates of the closest saltwater location.
 */
export class GetDistanceToWaterResponse {
    constructor(data = {}) {
        this.DistanceToWater = data.DistanceToWater;
        this.Latitude = data.Latitude;
        this.Longitude = data.Longitude;
        this.WaterLat = data.WaterLat;
        this.WaterLon = data.WaterLon;
        this.Error = data.Error ? new Error(data.Error) : null;
    }

    toString() {
        return `GetDistanceToWaterResponse: MilesToWater = ${this.DistanceToWater}, Latitude = ${this.Latitude}, Longitude = ${this.Longitude}, ClosestWaterLatitude = ${this.WaterLat}, ClosestWaterLongitude = ${this.WaterLon}, Error = ${this.Error ? this.Error.toString() : 'null'}`;
    }
}

/**
 * Response from GetReverseLocation API, containing the estimated address, city, state, ZIP code,
 * and county for the given coordinates.
 */
export class GetReverseLocationResponse {
    constructor(data = {}) {
        this.Address = data.Address;
        this.City = data.City;
        this.State = data.State;
        this.Zip = data.Zip;
        this.County = data.County;
        this.Error = data.Error ? new Error(data.Error) : null;
    }

    toString() {
        return `GetReverseLocationResponse: Address = ${this.Address}, City = ${this.City}, State = ${this.State}, Zip = ${this.Zip}, County = ${this.County}, Error = ${this.Error ? this.Error.toString() : 'null'}`;
    }
}

export default GetDistanceToWaterResponse;